Improve many things for our item parsing

kevr opened the pull request:
* Introduce more specific Item subclasses, so we can more easily
  produce json for specific cases, like a Map or a Fragment, without
  as many if/elif branches required.
* Improve get_item_modifiers_by_* functions in utils/trade.
  These now fetch ItemModifier objects directly instead of returning
  the entire dict containing them and their keys.
* Include map iiq/iir/pack_size when parsing an item. Currently,
  use a dict to decide if we have an "other_type" to return.
* Now support map influences.
* Support all item influences by putting them into the Item influence
  list.
* Support + and - for numerical modifiers. We can have items that have -n mods in them, so we include + and - now in the mod's value. This works fine through search and should through the API as well.

We still have some edge cases when parsing here that need to be worked
out. Please see the multi-line comment with a copied item and check
the current code against it. We miss a few explicit mods in that case.

There are also some augmented mods we're doing nothing with, but I'm
not sure that's much of an issue.

Additionally, there are many debugging print messages included with this commit
that need to be removed before we PR into upstream.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Improve many things for our item parsing

* Introduce more specific Item subclasses, so we can more easily
  produce json for specific cases, like a Map or a Fragment, without
  as many if/elif branches required.
* Improve get_item_modifiers_by_* functions in utils/trade.
  These now fetch ItemModifier objects directly instead of returning
  the entire dict containing them and their keys.
* Include map iiq/iir/pack_size when parsing an item. Currently,
  use a dict to decide if we have an "other_type" to return.
* Now support map influences.
* Support all item influences by putting them into the Item influence
  list.
* Additionally, replaced logging in the file's affected by the
  `logging` module. We can now enable DEBUG by passing -d or --debug
  to parse.py when starting it. Example: `sudo python3 parse.py --debug`

Caveats: See the 'reduced' to 'increased' conversion in the explicit
parsing section of this commit. We may need to do things like this for
other mods, though I have not yet encountered them as of this commit.

There are also some augmented mods we're doing nothing with, but I'm
not sure that's much of an issue.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr commented on parse.py:
```diff
@@ -43,22 +51,28 @@
 
 DEBUG = False
 
+MOD_TYPES = {
```
Should be removed. Not using this anymore.

kevr commented on parse.py:
```diff
@@ -43,22 +51,28 @@
 
+    # Remove the Note line; this is always located on the last line.
+    if item_list[-1][0].startswith("Note:"):
+        item_list = item_list[:-1]
```
Can probably in-place erase the last item instead of copying a slice.

kevr commented on parse.py:
```diff
@@ -67,13 +81,23 @@ def parse_item_info(text: str) -> Item:
+    iir = None
+    pack_size = None
+
+    item_class = Item
```
Default item class for now, we need to review how some of this stuff is organized, could probably use some better placement.

kevr commented on parse.py:
```diff
@@ -93,61 +118,152 @@ def parse_item_info(text: str) -> Item:
+                for e in item_list[1]
+            ]
+
+            # Parse out map-specific augmented fields.
```
All this stuff in this block can probably be done a bit better. Works for now, but should be revisited.

kevr commented on parse.py:
```diff
@@ -93,61 +118,152 @@ def parse_item_info(text: str) -> Item:
+
+            '''
+
+            if " (implicit)" in mod_text:
```
Probably a way to make this if/elif a bit more concise. I went with a hashtable but it just ended up being not as concise, and with a global so I didn't recreate it in loops. Needs work.w

kevr committed:
Properly close colors when logging

This was done in the previous commit. This is a text-only
commit detailing that this was done.

We should always immediately close colors when logging via
`Fore.RESET`, and avoid spaces being colored differently.

Additionally, Fore.RESET is better than using Fore.WHITE, as
it's more standard and will not force white text when the user's
terminal settings are different.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Introduce the ability to deduce items into their specific objects

We can now parse an item, which returns an Item type, then call
deduce_specific_object on that Item to convert it into it's
real subclass in O(n) where n = len(NINJA_BASES) time on first call,
and O(1) time on subsequent calls.

On my laptop, first call takes roughly 8.8 milliseconds to run, and
subsequent calls take 0.000001 seconds to run; a fraction of a
millisecond.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

kevr committed:
Remove timeit import

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>

sdf010 commented on models/Item.py:
```diff
@@ -1,31 +1,52 @@
+            logging.debug("%s: %s" % (k, v))
+        logging.debug("====== End of %s ======" % label)
+
+    def print_class(self):
```
why do we have print_class and print_attributes?

sdf010 replied:
This will be removed

sdf010 commented on parse.py:
```diff
@@ -35,30 +44,30 @@
     Parse item info (from clipboard, as obtained by pressing Ctrl+C hovering an item in-game).
     """
-    # TODO: test if poe item
+    # TODO: test if poe item (??)
```
Parse_item_info receives all text captured by the clipboard not just poe items. An idea is to move this parse function into the Item class as a class function and do the poe item check in the clipboard function.

sdf010 replied:
check len(item_list) > 2 and throw exception, catch in calling methods

sdf010 commented on models/Item.py:
```diff
@@ -1,31 +1,52 @@
-    name: str = 'Pseudo'
-    base: str = 'Pseudo'
+    name: str = None
+    base: str = None
```
If base is handled in the post init section we can remove it from here

sdf010 commented on parse.py:
```diff
@@ -93,61 +113,144 @@ def parse_item_info(text: str) -> Item:
         elif first_line.startswith('Item Level'):
             ilevel = first_line.lstrip('Item Level: ')
         elif first_line.count(' ') == 1 and first_line.endswith('Item'):
-            influence = []
```
Good spot here

sdf010 commented on parse.py:
```diff
@@ -93,61 +113,144 @@ def parse_item_info(text: str) -> Item:
+            logging.debug("Parsing %s" % line)
+            if " (implicit)" in mod_text:
+                item_type = ItemModifierType.IMPLICIT
+                mod_text = mod_text[:-11]
```
I prefer `mod_text = mod_text.rstrip(' (implicit)')` as it makes it obvious what is happening here instead of slices

sdf010 replied:
rstrip removes too much, so we will add comments instead

kevr replied:
Still haven't commented this in the most previous commit; apologies. I'll get to it asap here.

sdf010 commented on parse.py:
```diff
@@ -93,61 +113,144 @@ def parse_item_info(text: str) -> Item:
+            elif " (crafted)" in mod_text:
+                item_type = ItemModifierType.CRAFTED
+                logging.debug("Figuring out if '%s' is a crafted mod" % mod_text)
+                mod_text = mod_text[:-10]
```
same as above comment (for implicit)

sdf010 commented on parse.py:
```diff
@@ -93,61 +113,144 @@ def parse_item_info(text: str) -> Item:
-                        item_type = ItemModifierType.EXPLICIT
-                        mod = all_mods.get((mod_text, item_type))  # get a mod
-                        if mod is not None:
+                    elif "reduced" in mod_text:
```
should we do this earlier?

sdf010 replied:
We also need to do this for increased

sdf010 replied:
> should we do this earlier?

nope, we can not

sdf010 commented on parse.py:
```diff
@@ -928,66 +1041,65 @@ def price_item(text):
+        logging.info("====== ITEM DATA=====")
+        logging.info(f"{text}")
+        logging.info(f"{Fore.GREEN}================== END ISSUE DATA =================={Fore.RESET}")
+        logging.info(f"{Fore.RED}================== LOOKUP FAILED, PLEASE READ INSTRUCTIONS ABOVE =================={Fore.RESET}")
```
do we want to change this into a logging block with \n characters? it may look neater.

kevr replied:
Not concerned with this in these sets of commits. Something that should be taken up after the rewrite PR, or on upstream.

sdf010 reviewed (commented):
Some nice work here, I have a few questions which i have left as comments.

sdf010 commented on models/Item.py:
```diff
@@ -35,8 +56,207 @@ def __attrs_post_init__(self):
+class Amulet(Accessory):
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        super().print_attributes("Ring")
```
i'm sure this is meant to `super().print_attributes("Ring")`

kevr committed:
More progress on parse rework + new lookup + Item classes

With this commit, we can now lookup items with the old pricing
algorithm from the results we get + GUI generation.

Notable improvements:
    * Metamorph organs can now be priced
    * All item classes (with the exception of Beast) that exist can be
      looked up and priced.

Still missing:
    * parsing out Delve, Veiled, and other modifiers (see what we don't
      yet handle at the bottom of parse_item_info).
    * Overall cleanup; removing things we don't need anymore out of
      `parse.py`.
    * Pricing based on a uniform currency -- We currently sort through
      any results we find, regardless of currency, and make averages
      based on the numbers they give. We need to be converting currency
      from results we find to a uniform currency (chaos, exalted).
      This is something that is still an issue, afaik, in the current
      master/dev algorithm. We should improve this before the PR is
      actually created toward upstream.
    * Drop modifiers that are unimportant; example: reduced Attribute
      Requirements.
    * Relax the values that we use in the actual lookup json to produce
      more results. Currently, on items like 6L Loreweave, we match up mods
      so closely that we really do not get many results. This will
      require some serious consideration.
    * Quite a few more things... do note, that this is meant to be a
      partial update and is not meant to be something that is
      "finished."
    * Some more RestructuredText comments for classes and methods.

Signed-off-by: Kevin Morris <kevr.gtalk@gmail.com>
